### ISSUE TITLE
Keyboard layout stuck to English US

### ISSUE THREAD

**Issue by ovisergiu** [linuxmint/cinnamon#7792]
```
 * Cinnamon version (cinnamon --version) 3.8.8.
 * Distribution - (Mint 17.2, Arch, Fedora 25, etc...) Linux Mint 19 Cinnamon
 * Graphics hardware *and* driver used: Advanced Micro Devices, Inc. [AMD/ATI] Bonaire XTX [Radeon R7 260X/360] (prog-if 00 [VGA controller])
 * 64 bit
 ```

**Issue**
During Mint installtion I let keyboard layout to English US.
After successful installation of Mint 19, I add the Romanian keyboard layout. I put the Romanian layout in the top of the layout list but after every restart of Mint 19, the default layout is English US... Romanian is still there but is not the default.

**Steps to reproduce**

**Expected behaviour**

**Other information**

**Comment by icarter09** [comment]
Cinnamon 3.8.8
Linux Mint 19 64bit
Intel Corp HD Graphics 530

Verified that this is an issue. Issue #7606 seems to be very closely related, but that issue was closed out and working correctly now.

**Comment by ovisergiu** [comment]
Nope, the problems still exists... Today, when I first started the PC, the Romanian layout was default and without doing anything I just restarted the PC and the US was back again as default...

**Comment by icarter09** [comment]
@ovisergiu can you list the steps to reproduce this issue? I'm pretty sure I know them, but don't want to assume anything.

**Comment by ovisergiu** [comment]
Sure!
When installing Linux Mint 19, I let the English US but right after the Welcome screen appeared I openes Menu - Keyboards - Layouts tab - clicked the + sign - browsed for Romanian (standard), clic Add and in the list I could see English (US) being first and Romanian (standard) being second, but using the arrows from the same window I moved Romanian first and English to second. I also checked a box from the right part of the windows (Use a country flag, if available, to represent keyboards layouts). That was is. I restart the PC and noticed that the English flag was being displayed near the clock. When clicked the flag I could see the Romanian one too. Gave another restart but still English was the first to be displayed. Then I gave up, and I used the PC to watch movies and listen to music but watching carefully every time when I start Mint to see which flag will be displayed. Few times Romanian flag was displayed but most of the times, the English is displayed first. And of course, not just displayed, but the keyboard would type using the appropriate keys of the flag being displayed.

**Comment by lucky62** [comment]
I have the same problem. I have installed two layouts - english and slovak. English layout is always selected at start-up even the Slovak is selected as default (first in list). Keyboard applet is showing English flag.

![image](https://user-images.githubusercontent.com/16965519/52462225-77578000-2b72-11e9-8ca4-9e066d424382.png)

**Comment by ovisergiu** [comment]
I ”solved” this problem by reinstalling Mint and choosing Romanian as Default during installation and English as second. Good luck.

**Comment by MauroFoti** [comment]
> I ”solved” this problem by reinstalling Mint and choosing Romanian as Default during installation and English as second. Good luck.

I've also "solved" in another, less drastic way. I had the same problem with my Italian and English layout, and everytime the US would always be selected as startup. What I did was removing the Italian layout and adding it again immediately after, which solved the issue.

It looks like Cinnamon is somehow looking for layouts in cronological order, from the latest installed to the oldest, rather than following the order you give it.

**Comment by torchf1re** [comment]
I think cinnamon uses the Xorg keyboard configuration. There are multiple ways to set these configs. See:
https://wiki.archlinux.org/index.php/Xorg/Keyboard_configuration

Using localectl:
localectl set-x11-keymap <yourt desired layout>

**Comment by Vahan86** [comment]
@ovisergiu, is this still an issue in Cinnamon 4.2?

**Comment by ManIVIctorious** [comment]
I had exactly the same behaviour in cinnamon 4.2 in a virtual machine at university, it seems to be resolved since the release of 4.4 though.
At least for me, the issue does not exist anymore

**Comment by ovisergiu** [comment]
> @ovisergiu, is this still an issue in Cinnamon 4.2?

I am still having that problem.
I do not know how to update my Cinnamon version.
My Mint is updated daily but Cinnamon is still version 4.0.10

**Comment by milossramek** [comment]
Hi, I still have the problem on freshly installed mint 19.3 (cinnamon 4.4.8+tricia). It is sad, since I've just persuaded a colleague to switch from Windows.

**Comment by NicoG60** [comment]
I'm having the same issue with Mint 20 and Cinnamon 4.6.6

**Comment by NicoG60** [comment]
I tried to setup the US layout as well to have the little flag selector in the system tray. It seemed to work when I set it up but the keyboard layout applet was not loading after the next restart.
Furthermore, I tried to add the applet manually. When I clicked the plus button: The applets seemed to crash.
I'll have a deeper look at that crash and try to find a javascript stack dump that could help me/us track down that issue.
(If you have any idea where that can be found, that'd help. The Cinnamon looking glass doesn't say anything)

**Comment by NicoG60** [comment]
OK I've been looking into the issue.
Some context first, I've recently reinstalled my computer with Mint 20. My home directory is stored on another drive, so once the installation finished, I just edited fstab to mount the drive in the right place.

I think some of the configuration of the old cinnamon polluted the new installation. I found the ticket #8257 in which they discuss the keyboard applet crash I mentioned. (btw, I've seen the same error messages describe in that ticket).

The last comment of @camypaj describe what he did and unexpectedly resolve his issue (pasting it here for completeness):
> EDIT: what fixed it for me:
> 
> 1. in keyboard settings, click "reset to default"
> 2. right-click on a panel, and edit panel. Keyboard layout icon will appear, as a keyboard. Right-click it, and remove that applet.
> 3. log out, and log back in (that was enough for me, but I could try and reproduce it)
> 4. re-create the layouts
> 5. add the keyboard applet again.
> 
> Good news is that my original problem is now also resolved :)

I only did the 3 first step as I only need 1 keyboard layout. Just hitting that "Reset to default" button seemed to do the trick. I have no idea what happened under the hood. Probably it just reloaded an initial configuration stored in /etc and which was not corrupted by my fstab trick.

It's solved for me, hope it will help others.

**Comment by alextpedro** [comment]
I can confirm that this issue is still ongoing. The [steps posted by NicoG60](https://github.com/linuxmint/cinnamon/issues/7792#issuecomment-674406515) did not help. 
[System Info here](https://termbin.com/mn6gh).

**Comment by matlo** [comment]
I had the same issue on Linux Mint 20 cinnamon. I selected French layout on installation and added English US later (second entry in keyboard settings). Annoyingly English US kept being selected on startup. What fixed the issue for me is removing French in keyboard settings, **close keyboard settings**, reopen it, add French again and move it at the top.

**Comment by Chonnawonga** [comment]
Still getting this bug on 20.2 Cinnamon.

**Comment by sirnacnud** [comment]
This is still an issue on 21.1. I selected German layout on installation to match the physical keyboard. I added English US later and moved it to the top of the layout list to get priority. When after logging in, it always has the German layout selected.

**Comment by unalignedcoder** [comment]
Same here, same issue in Mint 21.1. There are three layouts, English is at the bottom but, no matter what layout is selected, at startup it always reverts back to English US. 
The interface language is English US, this may have something to do with it.

**Comment by ebourg** [comment]
> It may seem as if Mint will give priority to whatever layout matches the interface language

I'm not sure about this, my interface is in French but the layout still reverts to QWERTY.

The fun thing is, in the keyboard layout windows there are two French layouts configured and no English one.

**Comment by ebourg** [comment]
On my system `/etc/default/keyboard` is set to French, two French layouts are configured in the Cinnamon keyboard setup, and yet the keyboard is still English.

![Capture d’écran du 2023-07-20 18-46-40](https://github.com/linuxmint/cinnamon/assets/54304/7faa8246-0f47-4d31-ad76-8de915a2a541)

**Comment by wediaklup** [comment]
The issue seems to be still persistent. I wanted to swap from one german layout to another german layout and the original layout (chosen at OS installation) seems to always select itself.

**Comment by arielj** [comment]
This issue is still present in 2026. I only have a single keyboard distribution in the keyboard configuration: spanish latinamerica, but the layout keeps changing to english everytime I boot the machine, and I have to go to the distributions, add anything and then remove it so it picks up the spanish one

Is there any workaround at least?

**Comment by guigui2727** [comment]
I have the same kind of problem with Ubuntu Cinnamon 26.04 LTS.

After installing a new keyboard layout (using kalamine) and selected it on Cinnamon Keyboard settings, Cinnamon still fallback to the default en_us layout after a restart and loggin in. My keyboard layout is still there in Cinnamon Keyboard settings but even though, en_us layout is automatically used (it was the layout I used during my system installation}.

The workaround I found is to create a custom startup application (using "Startup Applications") with command : setxkbmap yourlayout -variant yourvariant  . For example, my command is: setxkbmap fr -variant bepolar  

This way, my keyboard layout is automatically launched at session startup.

Hope this help !
